Factory/DiactorosFactory.php

```diff
@@ -0,0 +1,159 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bridge\PsrHttpMessage\Factory;
+
+use Symfony\Bridge\PsrHttpMessage\HttpMessageFactoryInterface;
+use Symfony\Component\HttpFoundation\File\UploadedFile;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\StreamedResponse;
+use Zend\Diactoros\Response as DiactorosResponse;
+use Zend\Diactoros\ServerRequest;
+use Zend\Diactoros\ServerRequestFactory as DiactorosRequestFactory;
+use Zend\Diactoros\Stream as DiactorosStream;
+use Zend\Diactoros\UploadedFile as DiactorosUploadedFile;
+
+/**
+ * Builds Psr\HttpMessage instances using the Zend Diactoros implementation.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class DiactorosFactory implements HttpMessageFactoryInterface
+{
+    public function __construct()
+    {
+        if (!class_exists('Zend\Diactoros\ServerRequestFactory')) {
+            throw new \RuntimeException('Zend Diactoros must be installed to use the DiactorosFactory.');
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createRequest(Request $symfonyRequest)
+    {
+        $server = DiactorosRequestFactory::normalizeServer($symfonyRequest->server->all());
+        $headers = $symfonyRequest->headers->all();
+
+        try {
+            $body = new DiactorosStream($symfonyRequest->getContent(true));
+        } catch (\LogicException $e) {
+            $body = new DiactorosStream('php://temp', 'wb+');
+            $body->write($symfonyRequest->getContent());
+        }
+
+        $request = new ServerRequest(
+            $server,
+            DiactorosRequestFactory::normalizeFiles($this->getFiles($symfonyRequest->files->all())),
+            DiactorosRequestFactory::marshalUriFromServer($server, $headers),
```

Comment:
Because Symfony sometimes generate bad URI. For instance with the request I build in the test, the generated URL is `http://:/` (because some HTTP headers are missing).